### ISSUE TITLE
[post 1.7] Improve galaxy and globular rendering

### ISSUE THREAD

**Issue by SevenSpheres** [CelestiaProject/Celestia#544]
Note: this is also something for after the 1.7 release.

See #524, and celestia.Sci

**Comment by LukeCEL** [comment]
Maybe we should split this into several issues, since there are several problems with galaxy and globular rendering right now.

**Comment by SevenSpheres** [comment]
If you can list specific problems I can make a checklist and/or additional issues.

**Comment by LukeCEL** [comment]
GLOBULAR CLUSTERS:
- [ ] Individual sprites don't look very much like the stars in Celestia—they look too dim and dull.
- [ ] Some sprites look green, which shouldn't happen. It seems to depend on the KingConcentration parameter, and it's most obvious when KingConcentration ≈ 0.90. For an example, take a look at NGC 6712:
![ngc6712](https://user-images.githubusercontent.com/45546434/69891594-93780980-12cb-11ea-8dc7-b0ccbce8dd0c.png)
- [ ] There needs to be some way to make globular clusters elongated: #543 

I can add more later.

**Comment by LukeCEL** [comment]
- [ ] Globular clusters are too bright? #524 That being said, I can't seem to reproduce this on my build—they look fine to me.

GALAXIES:
- [ ] Elliptical galaxies don't look very similar to what they should look like. They should have a smooth, featureless profile and have a yellowish tone.
- [ ] Lenticular galaxies as well—they're quite diverse in appearance, but in general, they don't look as blue as spiral galaxies.
- [ ] Irregular galaxies need some way to make them elongated, sort of like E1-6 galaxies in 1.6.1. This applies specifically to galaxies like Wolf-Lundmark-Melotte.
- [ ] Specific to 1.7: Changing the light gain turns on and off sprites, so there's a weird twinkling lights effect.

**Comment by SevenSpheres** [comment]
@LukeCEL I put your checklist in the OP. @375gnu should I split these out into separate issues or leave them here?

**Comment by 375gnu** [comment]
Let's have them all together for a while.

> Some sprites look green, which shouldn't happen.

t00fri had an argue with @claurel regarding this issue you can try to find it on the forum.

**Comment by LukeCEL** [comment]
> t00fri had an argue with @claurel regarding this issue, you can try to find it on the forum. t00fri said this is correct colour.

Yeah, I saw that argument, but I couldn't really tell what he was saying. I think I might ask him on the Celestial Matters forums, but even the CM forums are having problems:
![image](https://user-images.githubusercontent.com/45546434/69904268-a2fe5d80-1372-11ea-85cd-ec03cc29d305.png)

**Comment by LukeCEL** [comment]
I asked again over at the CM forums:

> I've always noticed that in Celestia 1.6.1, globular clusters appear to have a ring of green stars around them, between the bright yellowish cores and the outer blue stars. Is this correct?

and he said something different:
> No, Luke. There are no green stars in visual-star color schemes. Since I did both the globulars in Celestia and those in celestia.Sci I remember that with the SDSS colors (my standard DSO reference), greenish shades might well appear. To compare stars in different schemes you must always fold in the color scheme used and then transform it to some standard reference scheme.

> Before coding the globulars I did many analytical studies via MAPLE. In celestia.Sci I further extended these studies and wrote a universal shader code for all stars (including halo stars around galaxies. globulars etc).

http://forum.celestialmatters.org/viewtopic.php?f=2&p=15516&sid=24f764d74b40e6a07334dc92bc73c738#p15516

**Comment by LukeCEL** [comment]
I tentatively fixed the green sprites problem by adding this code after line 689 of src/celengine/globular.cpp:

```
        if (hue > 60 && hue < 180) {
            sat = 0.0f;
        }
```

Should I make a PR for this?

**Comment by LukeCEL** [comment]
Here is a GIF demonstrating the "galaxy twinkling" effect:
![galaxytwinkling](https://user-images.githubusercontent.com/45546434/77112285-1e0d8180-69ff-11ea-8ddc-f60e6bfda671.gif)

**Comment by LukeCEL** [comment]
I think I've tracked down the commit that causes the "galaxy twinkling" bug: https://github.com/CelestiaProject/Celestia/commit/39eb715f984dc2fdd8e12d8bb7572a16c1695fba

**Comment by SevenSpheres** [comment]
Luke has been improving the rendering of elliptical and lenticular galaxies in [this thread](https://celestia.space/forum/viewtopic.php?f=6&t=20333). Could the concept behind this help with updating Celestia's code for rendering these types of galaxies?

**Comment by levinli303** [comment]
https://beltoforion.de/en/spiral_galaxy_renderer/ this looks good and it has code implementation.

**Comment by 375gnu** [comment]
> https://beltoforion.de/en/spiral_galaxy_renderer/ this looks good and it has code implementation.

this is was i was thinking about

**Comment by TheLostProbe** [comment]
we could try to recreate celestia.Sci's galaxy renderer, at least for spirals and irregulars, because there is enough information available to at least attempt this. Sci's galaxies look kind of similar to normal celestia's galaxies, but there is a colour gradient, the density of the nebulous parts is increased, the core has a bulge that glows (could be implemented in a similar way to proper elliptical galaxy rendering, essentially just a glowing yellow point in 3d), and instead of point sprites it uses diffusion limited aggregation. Sci's irregulars also use DLA, and there are two [entire](https://web.archive.org/web/20200803210835/http://forum.celestialmatters.org/viewtopic.php?f=11&t=771) [threads](https://web.archive.org/web/20200809204408/http://forum.celestialmatters.org/viewtopic.php?f=11&t=628) on the CM forums explaining it (sort of)
of course, the easier approach is going to be using the renderer on beltoforion.de, but it is going to need some changes since it is intended to be a dynamic, moving galaxy simulation, which is not compatible with celestia right now. but perhaps we should try something different first before resorting to that renderer

[three screenshots](https://cdn.discordapp.com/attachments/701749653657485373/1015134455440089189/unknown.png) [of Sci's](https://cdn.discordapp.com/attachments/701749653657485373/1015045968477044787/unknown.png) [spiral galaxies](https://cdn.discordapp.com/attachments/701749653657485373/1015045968896471110/unknown.png)


> Specific to 1.7: Changing the light gain turns on and off sprites, so there's a weird twinkling lights effect.

by the way, this should be ticked off because i dont think anyone has seen it in 1.7.0 for a long while now, including myself

**Comment by AstroChara** [comment]
For galaxies, I believe we might want to use volumetric objects to represent star glow, and for dusty galaxies we can perhaps repurpose the current particles to represent such dust clouds. The [Sérsic profile](https://en.wikipedia.org/wiki/Sérsic_profile) should be a very useful concept for implementing galaxies here. Would need to deproject it into 3D though.

**Comment by TheLostProbe** [comment]
Fridger made posts on the Celestial Matters forum regarding the use of the DLA algorithm to render irregular galaxies in celestia.Sci, it's possible this algorithm was used for spiral galaxies too. I think it's worth looking into this at some point.

https://web.archive.org/web/20160826114907/http://forum.celestialmatters.org/viewtopic.php?f=11&t=628&sid=6db4c7c02e352b14d808a3fe938eb58f

https://web.archive.org/web/20160826094920/http://forum.celestialmatters.org/viewtopic.php?f=11&t=771&sid=6db4c7c02e352b14d808a3fe938eb58f
